File: src/onion/compiler/env/java/ClassFileTable.java
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler.env.java;

import java.io.*;
import java.io.IOException;
import java.io.InputStream;

import org.apache.bcel.classfile.*;
import org.apache.bcel.classfile.ClassParser;
import org.apache.bcel.classfile.JavaClass;
import org.apache.bcel.util.Repository;
import org.apache.bcel.util.ClassPath;

/**
 * @author Kota Mizushima
 * Date: 2005/06/21
 */
public class ClassFileTable {
  private Repository repository;
  private ClassPath classPath;

  public ClassFileTable(String classPath) {
    this.classPath = new ClassPath(classPath);
    repository = org.apache.bcel.Repository.getRepository();
  }

  /**
   * @param className fully qualified class name
   * @return
   */
  public JavaClass load(String className) {
    try {
      return repository.loadClass(className);
    } catch (ClassNotFoundException e) {
      return add(className);
    }
  }

  private JavaClass add(String className){
    try {
      ClassPath.ClassFile classFile = classPath.getClassFile(className);
      InputStream input = classFile.getInputStream();
      String fileName = new File(classFile.getPath()).getName();
      ClassParser parser = new ClassParser(input, fileName);
      JavaClass javaClass = parser.parse();
      input.close();
      repository.storeClass(javaClass);
      return javaClass;
    } catch (IOException e) {
      return null;
    } catch (ClassFormatException e){
      return null;
    }
  }
}

File: src/main/scala/onion/compiler/environment/ClassFileTable.scala
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2016-, Kota Mizushima, All rights reserved.  *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler.environment

import java.io.{File, IOException}
import java.util.concurrent.ConcurrentHashMap

import onion.compiler.ExplicitClasspathClassLoader

/**
 * @author Kota Mizushima
 *
 */
class ClassFileTable(classPathString: String) {
  private val classLoader: ClassLoader = createClassLoader(classPathString)
  private val bytesCache = new ConcurrentHashMap[String, Option[Array[Byte]]]()

  private def createClassLoader(classPath: String): ClassLoader = {
    // `File.toURI` stats the path to decide on a trailing slash (directory vs jar), on
    // every entry of every compilation. The URL depends only on the path and that one
    // bit, so it is cached by both; the bit is still checked each time, so a directory
    // created after the first compilation is picked up.
    val urls =
      classPath.split(File.pathSeparator, -1).iterator
        .filter(_.nonEmpty)
        .map { path =>
          val file = new File(path)
          val key = if (file.isDirectory) path + File.separator else path
          ClassFileTable.urls.computeIfAbsent(key, _ => file.toURI.toURL)
        }
        .toArray
    ExplicitClasspathClassLoader(
      urls,
      Thread.currentThread().getContextClassLoader
    )
  }

  /**
   * Load class bytes for the given class name
   * @param className fully qualified class name
   * @return byte array of the class file, or null if not found
   */
  def loadBytes(className: String): Array[Byte] = {
    val cached = bytesCache.get(className)
    if (cached != null) return cached.orNull

    val resourcePath = className.replace('.', '/') + ".class"
    val inputStream = classLoader.getResourceAsStream(resourcePath)
    val loaded =
      if (inputStream == null) None
      else
        try {
          Option(inputStream.readAllBytes())
        } catch {
          case _: IOException => None
        } finally {
          inputStream.close()
        }

    bytesCache.putIfAbsent(className, loaded)
    loaded.orNull
  }
}

object ClassFileTable {
  private val urls = new ConcurrentHashMap[String, java.net.URL]()
}
